modules/frontend/querysharding.go

```diff
@@ -167,3 +154,41 @@ func DoRequests(ctx context.Context, downstream Handler, reqs []*http.Request) (
 
 	return resps, firstErr
 }
+
+func mergeResponses(rrs []RequestResponse) (*http.Response, error) {
+	var errCode = http.StatusOK
+	var errBody io.ReadCloser
+	var combinedTrace []byte
+	for _, rr := range rrs {
+		if rr.Response.StatusCode == http.StatusOK {
+			body, err := ioutil.ReadAll(rr.Response.Body)
+			if err != nil {
+				return nil, errors.Wrap(err, "error ready response body at query frontend")
+			}
+
+			if len(combinedTrace) == 0 {
+				combinedTrace = body
+			} else {
+				combinedTrace = util.CombineTraces(combinedTrace, body)
+			}
+		}
+		if rr.Response.StatusCode > errCode {
```

Comment:
I struggle with this.  Wouldn't we rather return 400 then 500?   If both are true we'd rather them retry thinking it's on their side since the 500 is probably transient (and may have been caused by the bad request).

@mdisibio any ideas for the best ways to combine http status codes?

Reply:
I feel like 5xx should always take precedence. From your other comment below, if we're return 5xx given one of the shards failed but another return 2xx, we should follow the same principle here for consistency right?

In the case below, I thought it would be good to atleast see a partial result from one of the successful shards even if some shards resulted in 5xx, but we can standardise on this.

Reply:
Agree with 5xx taking precedence.  It seems better to propagate errors clearly, or else it could be hard to detect issues and not realize partial trace data was returned.    Regarding transient errors:   yes some 5xx errors are retryable and sounds ok to require caller to retry.